pytest-embedded-jtag/pytest_embedded_jtag/openocd.py

```diff
@@ -1,36 +1,65 @@
 import logging
 import os
-from typing import Optional
+import shlex
+from typing import TYPE_CHECKING, Optional
 
 from pytest_embedded.log import DuplicateStdoutPopen, MessageQueue
 
+if TYPE_CHECKING:
+    from pytest_embedded_idf import IdfApp
+
 
 class OpenOcd(DuplicateStdoutPopen):
     """
     Class to communicate to OpenOCD
     """
 
     OPENOCD_PROG_PATH = 'openocd'
-    OPENOCD_DEFAULT_ARGS = '-f board/esp32-wrover-kit-3.3v.cfg -d2'
+    OPENOCD_DEFAULT_ARGS = '-f board/esp32-wrover-kit-3.3v.cfg'
 
-    TELNET_HOST = '127.0.0.1'
-    TELNET_PORT = 4444
+    TCL_BASE_PORT = 6666
+    TELNET_BASE_PORT = 4444
+    GDB_BASE_PORT = 3333
 
     def __init__(
         self,
         msg_queue: MessageQueue,
+        app: Optional['IdfApp'] = None,
         openocd_prog_path: Optional[str] = None,
         openocd_cli_args: Optional[str] = None,
+        port_offset: int = 0,
         **kwargs,
     ):
         openocd_prog_path = openocd_prog_path or os.getenv('OPENOCD_BIN', self.OPENOCD_PROG_PATH)
-        openocd_cli_args = openocd_cli_args or self.OPENOCD_DEFAULT_ARGS
+        openocd_cli_args = shlex.split(openocd_cli_args or self.OPENOCD_DEFAULT_ARGS)
 
         openocd_scripts_path = os.getenv('OPENOCD_SCRIPTS')
         if openocd_scripts_path:
-            openocd_cli_args += f' -s {openocd_scripts_path}'
+            openocd_cli_args.extend(['-s', openocd_scripts_path])
+
+        self.tcl_port = self.TCL_BASE_PORT + port_offset
+        self.telnet_port = self.TELNET_BASE_PORT + port_offset
+        self.gdb_port = self.GDB_BASE_PORT + port_offset
+
+        openocd_cli_args.extend(
+            [
+                '-c',
+                f'tcl_port {self.tcl_port}',
+                '-c',
+                f'telnet_port {self.telnet_port}',
+                '-c',
+                f'gdb_port {self.gdb_port}',
+            ]
+        )
+        if app and hasattr(app, 'flash_files'):  # IdfApp
+            # flash via jtag
+            for _f in app.flash_files:
+                if _f.encrypted:
+                    raise ValueError('Encrypted files can\'t be flashed in via JTAG')
+
+                openocd_cli_args.extend(['-c', f'program_esp {_f.file_path} {hex(_f.offset)} verify'])
```

Comment:
If there are multiple tests, will this happen on every test or just once per test app?

Reply:
cache added. thanks!